addons/point_of_sale/models/pos_payment_method.py

```diff
@@ -0,0 +1,114 @@
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+class PosPaymentMethod(models.Model):
+    """ Used to classify pos.payment.
+
+    Generic characteristics of a pos.payment is described in this model.
+    E.g. A cash payment can be described by a pos.payment.method with
+    fields: is_cash_count = True and a cash_journal_id set to an
+    `account.journal` (type='cash') record.
+
+    When a pos.payment.method is cash, cash_journal_id is required as
+    it will be the journal where the account.bank.statement.line records
+    will be created.
+    """
+
+    _name = "pos.payment.method"
+    _description = "Point of Sale Payment Methods"
+    _order = "id desc"
+
+    name = fields.Char(string="Name", required=True)
+    receivable_account_id = fields.Many2one('account.account',
+        string='Intermediary Account',
+        required=True,
+        domain=[('reconcile', '=', True), ('user_type_id.type', '=', 'receivable')],
+        default=lambda self: self.env.company.default_pos_receivable_account_id,
+        help='Account used as counterpart of the income account in the accounting entry representing the pos sales.')
+    is_cash_count = fields.Boolean(string='Cash')
+    cash_journal_id = fields.Many2one('account.journal',
+        string='Cash Journal',
+        domain=[('type', '=', 'cash')],
+        help='The payment method is of type cash. A cash statement will be automatically generated.')
+    split_transactions = fields.Boolean(
+        string='Split Transactions',
+        default=False,
+        help='If ticked, each payment will generate a separated journal item. Ticking that option will slow the closing of the PoS.')
+    session_ids = fields.Many2many('pos.session', string='Pos Sessions', help='Pos sessions that are using this payment method.')
+    company_id = fields.Many2one('res.company', string='Company', readonly=True, related='receivable_account_id.company_id')
+
+    @api.multi
+    def write(self, vals):
+        active_sessions = self.mapped('session_ids').filtered(lambda session: session.state != 'closed')
+        if active_sessions:
+            raise UserError('Modifying payment methods used in an open session is not allowed.\n'
+                            'Open sessions: %s' % (' '.join(active_sessions.mapped('name')),))
+        # set cash_journal_id to False if is_cash_count is changed to False
+        if not vals.get('is_cash_count', True):
+            vals['cash_journal_id'] = False
+        return super(PosPaymentMethod, self).write(vals)
+
+    @api.model
+    def generate_bank_and_cash_payment_methods(self):
```

Comment:
I'm not sure we need this method, as it will only add payment methods to the 'main shop', and it will just pre-generate payment methods links to journals but linked to nothing.

Reply:
I can't find it but Laura asked me to automatically create payment methods for bank and cash.

If I create them via .xml file, the cash_journal_id will be missed in the cash payment. At first, I implemented it by creating the payment methods during installation of chart of accounts. However, it ended up to be insufficient as account module can be installed before pos module. Few iterations led me to this solution. Can you recommend better solution?

Reply:
Indeed, Laura asked me to do so, to automatically create bank and cash payment methods for the main pos shop.